crates/primitives/src/block.rs

```diff
@@ -133,3 +137,271 @@ impl Decodable for BlockHashOrNumber {
         }
     }
 }
+
+#[derive(Copy, Clone, Debug, PartialEq, Eq)]
+/// A Block Identifier
+/// <https://github.com/ethereum/EIPs/blob/master/EIPS/eip-1898.md>
+pub enum BlockId {
+    /// A block hash
+    Hash(H256),
+    /// A block hash and a bool that defines if it's is canonical
+    HashReqCanonical((H256, bool)),
```

Comment:
This enum is a bit tricky, because ideally we want to have lossless serde. so we should keep the `bool` even though `Hash(h)` and `HashReqCanonical(h, false)`
would mean the same thing. 

Since we have a greenfield implementation and are not concerned with breaking things (unlike ethers) we can use create a new

`BlockHash` type 

```suggestion
BlockHash {
		/// hash of the block
		hash: H256,
		/// whether the block must be a canonical block
		require_canonical: Option<bool>,
	},
```

with some extra helper functions and `AsRef` traits, having this as standalone time makes this easier.

Reply:
Hi!
I'm finishing this in place of @xqft, what helper functions do you suggest adding? 
Nothing comes to mind for this.